### PR TITLE
AddLogEntry Overloads

### PR DESCRIPTION
**Pull request by LeeTwentyThree**
### Changes made in this pull request

  - Added `public static void AddLogEntry(string key, string languageKey, Sound sound, Sprite icon = null)`
  - Added `public static void AddLogEntry(string key, string languageKey, AudioClip audioClip, Sprite icon = null)`
  - Added `public static void AddLogEntry(string key, string languageKey, string soundFilePath, Sprite icon = null)`
  - Added `public static void AddLogEntry(string key, string languageKey, IFModSound fmodSoundInterface, Sprite icon = null)`

Can someone with more knowledge on FMOD (@Metious!) please review the logic of the added methods?

### CODE REVIEW

File: Nautilus/Handlers/PDAHandler.cs [hunk 1]
```diff
@@ -127,6 +128,56 @@ public static void AddLogEntry(string key, string languageKey, FMODAsset sound,
...
+    /// </summary>
+    /// <param name="key">The key (unique identifier) for this entry. Also used to create the sound asset, so make sure this string is <i>truly</i> unique!</param>
+    /// <param name="languageKey">The subtitles language key for this entry. Also see: <see cref="LanguageHandler.SetLanguageLine(string, string, string)"/>.</param>
+    /// <param name="sound">The sound that will be played once this log entry is unlocked or played through the PDA's Log tab.</param>
+    /// <param name="icon">The icon that will be used in the Log tab for this entry. if unassigned, it will use the default log entry icon.</param>
+    public static void AddLogEntry(string key, string languageKey, Sound sound, Sprite icon = null)
+    {
+        CustomSoundHandler.RegisterCustomSound(key, sound, AudioUtils.BusPaths.PDAVoice);
+        AddLogEntry(key, languageKey, AudioUtils.GetFmodAsset(key), icon);
+    }
```
Comment (Metious) [inline Nautilus/Handlers/PDAHandler.cs:142]: There are two problems with this code. 

Firstly, if there's already an existing custom sound registered with the same key, this method replaces it, but the previous sound will stay in memory causing a memory leak.

Secondly, the SoundQueue system (the one that's responsible for playing PDA sounds) uses the `id` part of the FMODAsset. Meaning in this case specifically, we have to pass an FMODAsset instance that sets the `id` field, not the `path`.

I suggest we change the logic so it only overrides the previous sound if they're not the same as the new one, if they aren't the same, we release the handle of the previous sound.
```suggestion
    public static void AddLogEntry(string key, string languageKey, Sound sound, Sprite icon = null)
    {
        /*
         * If there's an existing custom sound with the same key, we release the previous sound's handle.
         * This is so we do not leave any unused objects in memory.
         */
        if (CustomSoundHandler.TryGetCustomSound(key, out var prevSound))
        {
            if (prevSound.hasHandle() && prevSound.handle != sound.handle)
            {
                prevSound.release();
            }
        }
        
        CustomSoundHandler.RegisterCustomSound(key, sound, AudioUtils.BusPaths.PDAVoice);
        
        AddLogEntry(key, languageKey, AudioUtils.GetFmodAsset(key, key), icon);
    }
```

File: Nautilus/Handlers/PDAHandler.cs [hunk 2]
```diff
@@ -127,6 +128,56 @@ public static void AddLogEntry(string key, string languageKey, FMODAsset sound,
...
+    {
+        AddLogEntry(key, languageKey, CustomSoundHandler.RegisterCustomSound(key, soundFilePath, AudioUtils.BusPaths.PDAVoice), icon);
+    }
+
+    /// <summary>
+    /// Adds a custom PDA log entry message. Can be played by the Story Goal system or through <see cref="PDALog.Add"/>.
+    /// </summary>
+    /// <param name="key">The key (unique identifier) for this entry. Also used to create the sound asset, so make sure this string is <i>truly</i> unique!</param>
+    /// <param name="languageKey">The subtitles language key for this entry. Also see: <see cref="LanguageHandler.SetLanguageLine(string, string, string)"/>.</param>
+    /// <param name="fmodSoundInterface">The IFModSound interface that accessed when this log entry is played.</param>
```
Comment (Metious) [inline Nautilus/Handlers/PDAHandler.cs:173]: You do not pass interfaces on the overload, you pass an object that _implements_ the `IFModSound` interface.

I think it's more fitting if the argument was named `fmodSound` and the docs description was the same as the [RegisterCustomSound(string, IFModSound)](https://github.com/SubnauticaModding/Nautilus/blob/master/Nautilus/Handlers/CustomSoundHandler.cs#L76) overload.

**Review by Metious — CHANGES_REQUESTED** [review]
Added some comments. The rest looks good to me.

File: Nautilus/Handlers/PDAHandler.cs (hunk 1 above)
Comment (LeeTwentyThree) [reply Nautilus/Handlers/PDAHandler.cs:142]: `AudioUtils.GetFmodAsset(...)` already sets the ID and Path if the former is unassigned.

**Comment by LeeTwentyThree** [comment]
I left commits addressing everything you mentioned. However I was a bit unsure on one of them so I left an [experimental commit](https://github.com/SubnauticaModding/Nautilus/pull/15/commits/16df22030971009aa2b12afb1f4844fd7bb5ecbd) with quite a lengthy description.

**Comment by LeeTwentyThree** [comment]
Using your solution (reverted latest commit), but still not seeing the benefits of the Sound overload.

**Review by Metious — APPROVED** [review]

**Pull request merged**